master/buildbot/test/e2e/test_master_worker_create_start_stop.py

```diff
@@ -0,0 +1,921 @@
+    # worker.Worker and buildslave.BuildSlave are synonims, there is no actual
+    # need to use worker.Worker for buildbot-worker instances and
+    # buildslave.BuildSlave for deprecated buildslave instances.
+    worker.Worker('example-worker', 'pass'),
+    buildslave.BuildSlave('example-slave', 'pass'),
+]
+c['protocols'] = {'pb': {'port': 9989}}
+c['schedulers'] = [
+    schedulers.ForceScheduler(
+        name='worker-force',
+        builderNames=['worker-builder']),
+    schedulers.ForceScheduler(
+        name='slave-force',
+        builderNames=['slave-builder']),
+]
+
+worker_factory = util.BuildFactory(
+    [steps.ShellCommand(command=['echo', "Test worker"])])
+slave_factory = util.BuildFactory(
+    [steps.ShellCommand(command=['echo', "Test slave"])])
+
+c['builders'] = [
+    util.BuilderConfig(name='worker-builder', workernames=['example-worker'],
+        factory=worker_factory),
+    util.BuilderConfig(name='slave-builder', workernames=['example-slave'],
+        factory=slave_factory),
+]
+
+c['buildbotURL'] = 'http://localhost:8010/'
+c['www'] = dict(port=8010, plugins={})
+c['db'] = {
+    'db_url' : 'sqlite:///state.sqlite',
+}
+"""
+
+FILE_UPLOAD_TEST_CONFIG = """\
+from buildbot.plugins import *
+
+c = BuildmasterConfig = {}
+
+c['workers'] = [worker.Worker('example-worker', 'pass')]
+c['protocols'] = {'pb': {'port': 9989}}
+c['schedulers'] = [
+    schedulers.ForceScheduler(
+        name='force',
+        builderNames=['runtests'])
+]
+
+factory = util.BuildFactory()
+factory.addStep(
+    steps.StringDownload("filecontent", workerdest="dir/file1.txt"))
+factory.addStep(
+    steps.StringDownload("filecontent2", workerdest="dir/file2.txt"))
+factory.addStep(
+    steps.FileUpload(workersrc="dir/file2.txt", masterdest="master.txt"))
+factory.addStep(
+    steps.FileDownload(mastersrc="master.txt", workerdest="dir/file3.txt"))
+factory.addStep(steps.DirectoryUpload(workersrc="dir", masterdest="dir"))
+
+c['builders'] = [
+    util.BuilderConfig(name='runtests', workernames=['example-worker'],
+        factory=factory),
+]
+
+c['buildbotURL'] = 'http://localhost:8010/'
+c['www'] = dict(port=8010, plugins={})
+c['db'] = {
+    'db_url' : 'sqlite:///state.sqlite',
+}
+"""
+
+
+def wait_for_completion(is_completed_pred,
+                        check_interval=0.1, timeout=10, reactor=None):
+    if reactor is None:
+        reactor = global_reactor
+
+    @defer.inlineCallbacks
+    def next_try():
+        is_completed = yield is_completed_pred()
+
+        if is_completed:
+            caller.stop()
+        elif reactor.seconds() - start_time > timeout:
+            caller.stop()
+
+            raise RuntimeError("Timeout")
+
+    start_time = reactor.seconds()
+    caller = task.LoopingCall(next_try)
+    return caller.start(check_interval, now=True)
+
+
+def _read_dir_contents(dir):
+    contents = {}
+    for root, _, files in os.walk(dir):
+        for name in files:
+            fn = os.path.join(root, name)
```

Comment:
fn -> filename

(I'm curious what, for example, `pylint` has to say about this file...)